File: src/main/java/com/radbrackets/cinemachallenge/domain/Room.java
package com.radbrackets.cinemachallenge.domain;

import lombok.Value;

import java.time.ZonedDateTime;
import java.util.List;

@Value
class Room {
    String roomId;
    boolean isAvailable;
    List<ScheduledEvent> scheduledEvents;

    boolean incomingIsOverlapping(ScheduledEvent incomingEvent) {
        ZonedDateTime movieStart = incomingEvent.getEventStart();
        ZonedDateTime movieEnd = incomingEvent.getEventEnd();
        return scheduledEvents.stream().anyMatch(scheduledEvent -> {
            ZonedDateTime eventStart = scheduledEvent.getEventStart();
            ZonedDateTime eventEnd = scheduledEvent.getEventEnd();
            return movieStart.isAfter(eventStart) || movieEnd.isBefore(eventEnd);
        });
    }
}

File: src/main/kotlin/com/radbrackets/cinemachallenge/domain/Room.kt
package com.radbrackets.cinemachallenge.domain

import java.util.*

data class Room(
    val roomId: Int, val isAvailable: Boolean,
    val breakDurationInMinutes: Long, val seanses: MutableList<Seans> = Collections.synchronizedList(mutableListOf())
) {

    fun isSeansOverlapping(seans: Seans): Boolean {
        return seanses
            .any { seans.start >= it.start && seans.start <= it.end  || seans.end >= it.start && seans.end <= it.end }
    }
}